src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java

```diff
@@ -78,10 +77,17 @@ public static void addRoleToUser(Start start, String userId, String role)
 
     public static boolean createNewRoleOrDoNothingIfExists_Transaction(Start start, Connection con, String role)
             throws SQLException, StorageQueryException {
-        // We use IGNORE here since we want to ignore any conflict(duplicate key exception) that might be thrown.
-        // We do not use ON DUPLICATE UPDATE since it will return false positives when a role already exists
-        String QUERY = "INSERT IGNORE INTO " + Config.getConfig(start).getRolesTable() + " VALUES(?)";
-        int rowsUpdated = update(con, QUERY, pst -> pst.setString(1, role));
+        // We want to insert a role which does not exist into the roles table, Since INSERT queries do not support WHERE
+        // condition clause, we use a SELECT query with the input role and a WHERE NOT EXISTS clause to check that the
+        // input role does not exist in the table
+
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getRolesTable() + "(role) "
+                + "SELECT ? AS role WHERE NOT EXISTS (" + " SELECT role FROM " + Config.getConfig(start).getRolesTable()
+                + " WHERE role = ? )";
```

Comment:
Is this atmoic?

Reply:
So according to mysql docs [here](https://dev.mysql.com/doc/internals/en/transactions-historical-note.html), Mysql starts a transaction for every sql query, so the subqueries are also wrapped in a transaction. According to MySQL [docs](https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html) description of serializable, they convert regular SELECT to SELECT...FOR SHARE which would prevent the row from being deleted while this transaction is going on